gwnr/waveform/esigma_utils.py

```diff
@@ -443,23 +446,190 @@ def get_inspiral_esigma_waveform(
+        if hybridize_using_orbital_frequency:
+            window_start_idx = _get_window_start(
+                orbital_freq[: transition_idx + 1],
+                delta_t,
+                2 * num_hyb_orbits * np.pi,
+                direction="backward",
+            )
+        else:
+            # Orbital cycle based hybridization that keeps f_mr_transition
+            # at the hybridization frequency window's end
+            window_start_idx = (
+                np.searchsorted(
+                    orbital_phase,
+                    orbital_phase[transition_idx] - 2 * np.pi * num_hyb_orbits,
+                )
+                - 1
+            )
+        f_window_mr_transition = 2 * (
+            orbital_freq[transition_idx] - orbital_freq[window_start_idx]
+        )  # Extra 2-factor for returning the (2,2)-mode frequency
+
+        if verbose > 4:
+            print(
+                f"""The transition is to happen in a window from
+                frequencies: [{orbital_freq[window_start_idx]}, {orbital_freq[transition_idx]}]Hz,
+                between indices: [{window_start_idx}, {transition_idx}]"""
+            )
+
+    if keep_f_mr_transition_at_center:
+        # Orbital cycle based hybridization that keeps f_mr_transition at
+        # the hybridization frequency window's midpoint
+        if (orbital_phase[transition_idx] + num_hyb_orbits * np.pi) > orbital_phase[-1]:
+            raise RuntimeError(
+                f"""Requested number of orbits to hybridize over not available
+in the waveform after the transition frequency. Either decrease the number of
+orbits to hybridize over (currently {num_hyb_orbits}) or decrease the
+inspiral-to-merger transition frequency."""
+            )
+
+        if hybridize_using_orbital_frequency:
+            window_start_idx = _get_window_start(
+                orbital_freq[: transition_idx + 1],
+                delta_t,
+                num_hyb_orbits * np.pi,
+                direction="backward",
+            )
+            window_end_idx = transition_idx + _get_window_start(
+                orbital_freq[transition_idx:],
+                delta_t,
+                num_hyb_orbits * np.pi,
+                direction="forward",
+            )
+        else:
+            # phase is always a monotonically increasing function,
+            #    hence using the more efficient np.searchsorted method
+            window_start_idx = -1 + np.searchsorted(
+                orbital_phase, orbital_phase[transition_idx] - num_hyb_orbits * np.pi
+            )
+            window_end_idx = np.searchsorted(
+                orbital_phase, orbital_phase[transition_idx] + num_hyb_orbits * np.pi
+            )
+        f_window_mr_transition = 4 * np.max(
+            [
+                orbital_freq[window_end_idx] - f_mr_transition / 2.0,
+                f_mr_transition / 2.0 - orbital_freq[window_start_idx],
+            ]
+        )  # Extra 2-factor for returning the (2,2)-mode frequency
+
```

Comment:
Simplification of the above conditions:

```
    if keep_f_mr_transition_at_center:
        # Orbital cycle based hybridization that keeps f_mr_transition at
        # the hybridization frequency window's midpoint
        if (orbital_phase[transition_idx] + num_hyb_orbits * np.pi) > orbital_phase[-1]:
            raise RuntimeError(
                f"""Requested number of orbits to hybridize over not available
in the waveform after the transition frequency. Either decrease the number of
orbits to hybridize over (currently {num_hyb_orbits}) or decrease the
inspiral-to-merger transition frequency."""
            )

        if hybridize_using_orbital_frequency:
            window_start_idx = _get_window_start(
                orbital_freq[: transition_idx + 1],
                delta_t,
                num_hyb_orbits * np.pi,
                direction="backward",
            )
            window_end_idx = transition_idx + _get_window_start(
                orbital_freq[transition_idx:],
                delta_t,
                num_hyb_orbits * np.pi,
                direction="forward",
            )
        else:
            # phase is always a monotonically increasing function,
            #    hence using the more efficient np.searchsorted method
            window_start_idx = -1 + np.searchsorted(
                orbital_phase, orbital_phase[transition_idx] - num_hyb_orbits * np.pi
            )
            window_end_idx = np.searchsorted(
                orbital_phase, orbital_phase[transition_idx] + num_hyb_orbits * np.pi
            )
        f_window_mr_transition = 4 * np.max(
            [
                orbital_freq[window_end_idx] - f_mr_transition / 2.0,
                f_mr_transition / 2.0 - orbital_freq[window_start_idx],
            ]
        )  # Extra 2-factor for returning the (2,2)-mode frequency
   
    else:
        if hybridize_using_orbital_frequency:
            window_start_idx = _get_window_start(
                orbital_freq[: transition_idx + 1],
                delta_t,
                2 * num_hyb_orbits * np.pi,
                direction="backward",
            )
        else:
            # Orbital cycle based hybridization that keeps f_mr_transition
            # at the hybridization frequency window's end
            window_start_idx = (
                np.searchsorted(
                    orbital_phase,
                    orbital_phase[transition_idx] - 2 * np.pi * num_hyb_orbits,
                )
                - 1
            )
        f_window_mr_transition = 2 * (
            orbital_freq[transition_idx] - orbital_freq[window_start_idx]
        )  # Extra 2-factor for returning the (2,2)-mode frequency

        if verbose > 4:
            print(
                f"""The transition is to happen in a window from
                frequencies: [{orbital_freq[window_start_idx]}, {orbital_freq[transition_idx]}]Hz,
                between indices: [{window_start_idx}, {transition_idx}]"""
            )
```

This avoids `failsafe` altogether (hence it can be removed from this function), saves some unnecessary checks of the condition `orbital_phase[transition_idx] + num_hyb_orbits * np.pi) > orbital_phase[-1]` and saves the unnecessary hybridization window computation for the case when `keep_f_mr_transition_at_center`, `failsafe` and `orbital_phase[transition_idx] + num_hyb_orbits * np.pi) > orbital_phase[-1]` are all True, in which case the RunTimeError will be executed anyway. 

Also regarding this last condition, it means that even if `failsafe=True` and  `orbital_phase[transition_idx] + num_hyb_orbits * np.pi) > orbital_phase[-1]`, and if the user demands `keep_f_mr_transition_at_center`, we are not switching to the scheme where we keep the transition frequency at the edge. It seems fine to me as switching from centre to edge would not be good for parameter-sapce continuity of the waveform as well. But just highlighting this case for your comments

Reply:
I am not sure I understand this, this seems to me the same exact code logic with the if-else cases reversed in order + the case for `failsafe=True` eliminated entirely. 

Lets discuss this in person?

Reply:
If `failsafe` is activated, it assumes we want to now keep the alignment window's `f_mr_transition` at the end of the window, i.e. it automatically disables `keep_f_mr_transition_at_center`. This is because `f_mr_transition` is set to `mode_frequency.max()`, which could very well be at the last point of inspiral frequency evolution.

Reply:
Remember that if `failsafe` is activated because `mode_frequency.max() < f_mr_transition` then the model will not be continuous across the parameter space anyway. We have literally no way of avoiding that.

Reply:
And the code above automatically activates the `not keep_f_mr_transition_at_center` in `failsafe` mode when its unable to keep that frequency at center since `(orbital_phase[transition_idx] + num_hyb_orbits * np.pi) > orbital_phase[-1]`.

This seems like reasonable `failsafe` behavior. We can also shrink the `f_mr_transition` centered window, that is another option.

Reply:
Just pushed a fixup for this @Akash-Maurya-0899 , as we discussed today.

Reply:
can we resolve this @Akash-Maurya-0899 ?